src/TSne.jl

```diff
@@ -281,4 +263,48 @@ function tsne(X::Matrix, ndims::Integer = 2, reduce_dims::Integer = 0,
     return Y
 end
 
+function computeY(X::Matrix, pca_init::Bool, ndims::Integer, verbose::Bool=true)
+    (n, d) = size(X)
+    if !pca_init
+        verbose && info("Starting with random layout...")
+        Y = randn(n, ndims)
+    else
+        verbose && info("Using the first $ndims components of the data PCA as the initial layout...")
+        if reduce_dims >= ndims
+            Y = X[:, 1:ndims] # reuse X PCA
+        else
+            @assert reduce_dims <= 0 # no X PCA
+            Y = pca(X, ndims)
+        end
+    end
+    Y
+end
+
+"Tsne where `X` is a matrix of values and distance is Euclidean"
+function tsne(X::Matrix; ndims::Integer=2, reduce_dims::Integer = 0, verbose::Bool=true,
```

Comment:
You can easily merge this `tsne()` function and the function below, just declare `X::Union{Matrix, Vector}` and make your `δ` a keyword argument (`distance::Union{Function, Void}`, defaults to `nothing`).
Then you may implement three `pairwisedists` methods:
* `pairwisedists(X::Matrix, ::Void)` default Euclidean distances
* `pairwisedists(X::Vector, dist::Function)`
* `pairwisedists(X::Matrix, dist::Function)` distances between the `X` rows using `dist` function

The call `pairwisedists(X, dist)` would automatically dispatch to the right method.

Reply:
What is missing so far is the possibility to specify that `X` is the distance matrix itself.
One possibility is to allow `distance` keyword argument to be boolean instead of void:
 * `distance::Function` specifies that `X` contains objects and `distance` is a function to calculate the distance between a pair of objects
 * `distance = false` specifies that `X` contains objects and pairwise distances should be calculated using Euclidean metric
 * `distance = true` specifies that `X` is a matrix of pairwise distances (in that case you should check that it's non-negative and symmetrical)